### PR TITLE
Support test files with same name in different dirs

### PR DESCRIPTION
**Pull request by taiki-e**
Currently, macrotest does not support test files with the same name in different dirs.

Repro:

```sh
cargo new --lib repro
cd repro
mkdir tests
mkdir tests/expand
mkdir tests/expand/a
mkdir tests/expand/b
echo 'macrotest = "1"' >> Cargo.toml
echo 'fn main() {}' > tests/expand/a/t.rs
echo 'fn main() {}' > tests/expand/b/t.rs
echo '#[test] fn expand() { macrotest::expand("tests/expand/*/*.rs") }' > tests/expandtest.rs
cargo test
```

Result:

```text
error: failed to parse manifest at `/Users/taiki/tmp/repro/target/tests/repro/mB6czU8BSkq7oGl4v9vaSYNOytDoCUaHoONbPeBDUF/Cargo.toml`
Caused by:
  found duplicate binary name t, but all binary targets must have a unique name
test expand ... FAILED

failures:

---- expand stdout ----
Running 2 macro expansion tests
Expansion error:
error: failed to parse manifest at `/Users/taiki/tmp/repro/target/tests/repro/mB6czU8BSkq7oGl4v9vaSYNOytDoCUaHoONbPeBDUF/Cargo.toml`
Caused by:
  found duplicate binary name t, but all binary targets must have a unique name

Expansion error:
error: failed to parse manifest at `/Users/taiki/tmp/repro/target/tests/repro/mB6czU8BSkq7oGl4v9vaSYNOytDoCUaHoONbPeBDUF/Cargo.toml`
Caused by:
  found duplicate binary name t, but all binary targets must have a unique name
```

This is because macrotest determines the test name based on the filename. The generated Cargo.toml contains bins with name conflicts such as:

```
[[bin]]
name = "t"
path = "/Users/taiki/tmp/repro/tests/expand/a/t.rs"

[[bin]]
name = "t"
path = "/Users/taiki/tmp/repro/tests/expand/b/t.rs"
```

This patch fixes the problem by giving bins a unique name. This is basically the same method that trybuild actually uses. (https://github.com/dtolnay/trybuild/commit/aa544645e9de881d2ca1dabe92984e56618eb7ca)

### CODE REVIEW

File: src/expand.rs [hunk 1]
```diff
@@ -128,29 +128,22 @@ where
...
-            .into_owned();
+        let path = test.test.display();
+        let expanded_path = test.test.with_extension("expanded.rs");
 
         match test.run(&project, expansion_behavior, &args) {
             Ok(outcome) => match outcome {
-                ExpansionOutcome::Same => println!("{} - ok", file_stem),
+                ExpansionOutcome::Same => {
+                    let _ = writeln!(std::io::stdout(), "{} - ok", path);
+                }
```
Comment (taiki-e) [inline src/expand.rs:138]: This change isn't directly related to this fix, but `cargo test` doesn't print the output from `println!`, so this is needed to make sure the path was output correctly.

File: src/expand.rs [hunk 2]
```diff
@@ -128,29 +128,22 @@ where
...
     let mut failures = 0;
     for test in tests {
-        let file_stem = test
-            .test
-            .file_stem()
-            .expect("no file stem")
-            .to_string_lossy()
-            .into_owned();
+        let path = test.test.display();
+        let expanded_path = test.test.with_extension("expanded.rs");
```
Comment (eupn) [inline src/expand.rs:132]: Can you tweak the `EXPANDED_RS_SUFFIX` constant to have this value (without the starting dot) and then put the constant here instead?

```suggestion
        let expanded_path = test.test.with_extension(EXPANDED_RS_SUFFIX);
```

File: src/expand.rs (hunk 1 above)
Comment (eupn) [reply src/expand.rs:138]: Right. It prints it only when something fails, for example, when one of the expansions failed, you can see which ones were ok.

File: src/expand.rs (hunk 2 above)
Comment (eupn) [reply src/expand.rs:132]: I like this particular change a lot, it's much cleaner!

**Comment by eupn** [comment]
@taiki-e could you please also add a regression test for this? You can modify the existing test macro projects to reflect this edge case or create a separate test macro project for this.

**Comment by taiki-e** [comment]
@eupn Added a regression test and replaced `"expanded.rs"` with `EXPANDED_RS_SUFFIX`.

**Review by eupn — APPROVED** [review]
Excellent, @taiki-e. Thank you for covering this. Let's wait for CI and then merge.
